nomos-mix/message/src/keyset.rs

```diff
@@ -0,0 +1,253 @@
+    ///
+    /// Example:
+    /// As a packet builder:
+    ///     shared1 = DH(pub1, epriv0)
+    ///     keyset1 = KeySet(shared1)
+    ///     shared2 = DH(DH(pub2, epriv0), keyset1.epriv)
+    ///     keyset2 = KeySet(shared2)
+    /// As the 1st recipient:
+    ///     shared1 = DH(epub0, priv1)
+    ///     keyset1 = KeySet(shared1)
+    ///     epub1 = DH(epub0, keyset1.epriv)
+    /// As the 2nd recipient:
+    ///     shared2 = DH(priv2, epub0) = DH(priv2, DH(epub0, keyset1.epriv))
+    ///     keyset2 = KeySet(shared2)
+    ///
+    pub fn derive_all(
+        recipient_pubkeys: &[x25519_dalek::PublicKey],
+        initial_ephemeral_privkey: &x25519_dalek::StaticSecret,
+    ) -> Vec<KeySet> {
+        let mut keysets = Vec::with_capacity(recipient_pubkeys.len());
+
+        let mut ephemeral_privkeys = vec![initial_ephemeral_privkey.clone()];
+
+        // For each recipient, derive a shared secret by accumulated DH operations
+        // and derive a [`KeySet`] from the shared secret.
+        recipient_pubkeys
+            .iter()
+            .cloned()
+            .for_each(|recipient_pubkey| {
+                // Derive a shared secret of the recipient by accumulated DH operations.
+                let shared_secret = ephemeral_privkeys
+                    .iter()
+                    .fold(recipient_pubkey, |accumulated_pubkey, ephemeral_privkey| {
+                        let shared_secret = ephemeral_privkey.diffie_hellman(&accumulated_pubkey);
+                        // Convert SharedSecret to PublicKey just for the next DH operation.
+                        // This is just converting the type, not manipulating bytes.
+                        x25519_dalek::PublicKey::from(shared_secret.to_bytes())
+                    })
+                    .to_bytes(); // Convert the final PublicKey back to the bytes for SharedSecret.
+
+                // Derive a [`KeySet`] from the shared secret
+                // and keep the derived new ephemeral private key for the next recipient.
+                let keyset = KeySet::derive(&shared_secret);
+                ephemeral_privkeys.push(keyset.ephemeral_private_key.clone());
+                keysets.push(keyset);
+            });
+
+        keysets
+    }
+
+    /// Derive a ephemeral public key to be shared to the next recipient
+    /// by the Diffie-Hellman operation using the shared public key
+    /// and the derived ephemeral private key of the current recipient.
+    ///
+    /// This is the operation symmetric to the one of accumulated DH operations
+    /// performed in the [`KeySet::derive_all`].
+    pub fn derive_next_ephemeral_public_key(
+        &self,
+        shared_ephemeral_pubkey: &x25519_dalek::PublicKey,
+    ) -> x25519_dalek::PublicKey {
+        let new_shared_secert = self
+            .ephemeral_private_key
+            .diffie_hellman(shared_ephemeral_pubkey);
+        x25519_dalek::PublicKey::from(new_shared_secert.to_bytes())
+    }
+}
```

Comment:
Cant we use all this from the original sphix packet crate? My point is if we do not modify somehting and it can be reused is better. Easier for maintain.

Reply:
Let me check. Their code is not modularized well. and I'm not sure if they provide useful public functions.

Reply:
I just tried to use functions defined in https://github.com/nymtech/sphinx: 7e07e8488b68e500ef4f3ae8ecf3aeab25e826df in the [`mix-msg-using-sphinx`](https://github.com/logos-co/nomos-node/tree/mix-msg-using-sphinx) branch. I could remove `keyset.rs` and `payload.rs` by reusing their functions. But, unfortunately, only one function that is necessary for deriving the ephemeral public key for the next recipient is not exposed by the sphinx library: [`blind_the_shared_secret`](https://github.com/nymtech/sphinx/blob/344b902df340e0d5af69c5147b05f76f324b8cef/src/header/mod.rs#L234). So, I copied the function to our codebase: [`derive_next_ephemeral_public_key`](https://github.com/logos-co/nomos-node/blob/7e07e8488b68e500ef4f3ae8ecf3aeab25e826df/nomos-mix/message/src/packet.rs#L117).

That function is tightly coupled with the header implementation in Sphinx. However, their header implementation contains some metadata that isn't necessary to us. Also, we're going to add some new metadata to the header that is going to be defined by us (e.g. incentivization). That's way we cannot use their header implementation.

Hm, let me think. The easiest way may be just using the Sphinx as it is for now, without modifying anything, with filling dummy data into unnecessary metadata. It will work. But, in the end, I think we need to adopt all details from them and modify them on our own.

Reply:
Well, we can just modify that part. If we cannot reuse it its fine to have a copy of it ourselves. But if we reduced the duplication and maintain of code we do not really own is better. Thanks for the extra effort!!

Reply:
Okay. I merged the `mix-msg-using-sphinx` branch into this branch with some refactoring. You can see the change 8cb0bc7fb3d65791faeaf99b6e0f94814d3c2975. As mentioned above, I copied one function from their code. Also you may find that there is a "not-natural" piece of code regarding "dummy node address", even though we don't need any node address in our design. It's because their code isn't modularized well. I left explanation in code comment.

So far, it's fine. However, we won't be able to reuse their code for encrypting headers (which I'll do at the next PR). Their code is tightly coupled with their own header definition and its size. So, the next PR will contain some complicated algorithms modified from their code. I'll try my best to reuse pieces from their code, but it's not gonna be as simple as this PR.